File: avin_data/src/manager/source.py
#!/usr/bin/env  python3
# ============================================================================
# URL:          http://avin.info
# AUTHOR:       Alex Avin
# E-MAIL:       mr.alexavin@gmail.com
# LICENSE:      MIT
# ============================================================================

from __future__ import annotations

import enum

from src.exceptions import NotImplemetedSource


class Source(enum.Enum):
    """Market data source."""

    MOEX = 1
    TINKOFF = 2

    @classmethod
    def from_str(cls, string: str) -> Source:
        """Get enum from str.

        Args:
            string: category name.

        Returns:
            Category Enum.

        Raises:
            NotImplemetedSource if category not exists.
        """
        if attr := getattr(cls, string.upper(), None):
            return attr
        raise NotImplemetedSource(
            f"Source not implemented. Choice from {cls.__members__.keys()}"
        )


if __name__ == "__main__":
    ...

File: avin_data/src/source.rs
/*****************************************************************************
 * URL:         http://avin.info
 * AUTHOR:      Alex Avin
 * E-MAIL:      mr.alexavin@gmail.com
 * LICENSE:     MIT
 ****************************************************************************/

/// List for selecting the source of downloading market data.
///
/// # ru
/// Перечисление для выбора источника загрузки рыночных данных.
#[derive(Debug, PartialEq, Eq, Clone, Copy, Hash, strum::Display)]
pub enum Source {
    MOEX,
    TINKOFF,
}
impl Source {
    /// Return market data source name.
    ///
    /// # ru
    /// Возвращает название источника биржевых данных.
    pub fn name(&self) -> &'static str {
        match self {
            Self::MOEX => "MOEX",
            Self::TINKOFF => "TINKOFF",
        }
    }
}
impl From<&str> for Source {
    fn from(value: &str) -> Source {
        match value.to_uppercase().as_str() {
            "MOEX" => Source::MOEX,
            "TINKOFF" => Source::TINKOFF,
            _ => panic!("Invalid source: {value}"),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn name() {
        assert_eq!(Source::MOEX.name(), "MOEX");
        assert_eq!(Source::TINKOFF.name(), "TINKOFF");
    }
    #[test]
    fn to_str() {
        assert_eq!(Source::MOEX.to_string(), "MOEX");
        assert_eq!(Source::TINKOFF.to_string(), "TINKOFF");
    }
    #[test]
    fn from_str() {
        assert_eq!(Source::MOEX, "moex".into());
        assert_eq!(Source::TINKOFF, "TiNkoFf".into());
    }
}
